### PR TITLE
distillation v0

### PR DESCRIPTION
**Pull request by Bezdarnost**
Now we can provide path to the PT file with logprobs in training. 

Ignore mixin.py changes.

### CODE REVIEW

File: swift/trainers/trainers.py [hunk 1]
```diff
@@ -189,7 +189,7 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
                 model_name = unwrapped_model._get_name()
             # User-defined compute_loss function
             if self.compute_loss_func is not None:
-                loss = self.compute_loss_func(outputs, labels, num_items_in_batch=num_items_in_batch, **loss_kwargs)
+                loss = self.compute_loss_func(outputs, labels, num_items_in_batch=num_items_in_batch, , logprobs_path=inputs.pop('logprobs_path'), **loss_kwargs)
```
Comment (young-agi) [inline swift/trainers/trainers.py:192]: Here you have an extra `,` - need to remove it.

File: swift/plugin/loss.py [hunk 2]
```diff
@@ -382,6 +383,70 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
     return loss
 
 
+@register_loss_func(LossType.logits_distillation)
+def logits_distillation_loss(outputs, labels, loss_scale=None, num_items_in_batch=None, logprobs_path=None) -> torch.Tensor:
```
Comment (young-agi) [inline swift/plugin/loss.py:387]: Ideally the logprobs must be loaded in the dataloader, when the batch is prepared, and not during the los calculation. This will make things slower.

**Review by young-agi — CHANGES_REQUESTED** [review]
Let's move logprob loading to Dataloder after you finish the debugging of the loss function.

File: swift/plugin/loss.py [hunk 3]
```diff
@@ -382,6 +383,70 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
...
+    alpha        = 0.5 # Weight between CE and KL, 1 - without CE, 0 - without KL
+    temperature  = 1.0
+    eps          = 1e-12
+
+    logits  = outputs.logits                               # [B,L,V]
+    device  = logits.device
+    B       = logits.size(0)
+
+    # ----- teacher top-k data ------------------------------------------------
+    if token_indices is None or token_logprobs is None:
```
Comment (young-agi) [inline swift/plugin/loss.py:397]: token_indices and token_logprobs are not defined!

Comment (Bezdarnost) [reply swift/plugin/loss.py:397]: yes, I removed it and didn't change the code, my bad

File: swift/plugin/loss.py [hunk 4]
```diff
@@ -382,6 +383,70 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
...
+        assert logprobs_path is not None, \
+            "Need `logprobs_path` if token_indices/logprobs aren’t provided."
+        idx_list, lp_list = [], []
+        for p in logprobs_path:               # each file: (k, L_i)
```
Comment (young-agi) [inline swift/plugin/loss.py:401]: NIT: can you please write more meaningful comments. Eg.
 in shape `(k, L_i)` what is k and what us L_i ?

File: swift/plugin/loss.py (hunk 2 above)
Comment (Bezdarnost) [reply swift/plugin/loss.py:387]: Thank you, I will move it in the Dataset class

File: swift/trainers/trainers.py (hunk 1 above)
Comment (Bezdarnost) [reply swift/trainers/trainers.py:192]: Thank you, I didn't notice, I will fix

File: swift/plugin/loss.py [hunk 5]
```diff
@@ -382,6 +383,70 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
...
+            idxs, lps = torch.load(p)
+            idx_list.append(idxs.T)           # (L_i, k)
+            lp_list.append(lps.T)
+        token_indices  = _pad_stack(idx_list,  pad_value=0)
```
Comment (young-agi) [inline swift/plugin/loss.py:405]: `_pad_stack` is not defined

File: swift/plugin/loss.py (hunk 4 above)
Comment (Bezdarnost) [reply swift/plugin/loss.py:401]: Okay, I will remember

File: swift/plugin/loss.py [hunk 6]
```diff
@@ -382,6 +383,70 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
...
+        token_logprobs = _pad_stack(lp_list, pad_value=-1e9)
+
+    token_indices  = token_indices.to(device).long()       # ← cast to int64
+    token_logprobs = token_logprobs.to(device)
+
+    # ----- teacher distribution ---------------------------------------------
+    teacher_logp = token_logprobs / temperature            # [B,L,k]
+    teacher_p    = teacher_logp.exp()
+    teacher_p   /= teacher_p.sum(-1, keepdim=True).clamp_min(eps)
+
```
Comment (young-agi) [inline swift/plugin/loss.py:420]: Better use softmax function from PyTorch for efficiency. You implementations might be not efficient

File: swift/plugin/loss.py (hunk 5 above)
Comment (Bezdarnost) [reply swift/plugin/loss.py:405]: added

File: swift/plugin/loss.py (hunk 6 above)
Comment (Bezdarnost) [reply swift/plugin/loss.py:420]: will change

**Comment by Bezdarnost** [comment]
TODO:
- rewrite softmax with softmax from F
- move reading of logprobs in dataset

File: swift/plugin/loss.py [hunk 7]
```diff
@@ -381,6 +382,76 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
...
+    # ----- teacher top-k data ------------------------------------------------
+    idx_list, lp_list = [], []
+    for p in logprobs_path:               # each file: (k, L_i)
+        idxs, lps = torch.load(p)
+        idx_list.append(idxs.T)           # (L_i, k)
+        lp_list.append(lps.T)
+    token_indices  = _pad_stack(idx_list,  pad_value=0)
+    token_logprobs = _pad_stack(lp_list, pad_value=-1e9)
+
+    token_indices  = token_indices.to(device).long()       # ← cast to int64
```
Comment (young-agi) [inline swift/plugin/loss.py:419]: to avoid extra conversion to `long`, one could pass the `torch.int64` dtype in _pad_stack() call

Comment (young-agi) [reply swift/plugin/loss.py:419]: Do we really need int64 btw ? int32 should be enough

File: swift/plugin/loss.py (hunk 6 above)
Comment (young-agi) [reply swift/plugin/loss.py:426]: Just use this line instead
`teacher_probs = torch.softmax(token_logprobs / temperature, dim=-1)`

File: swift/plugin/loss.py [hunk 8]
```diff
@@ -381,6 +382,76 @@ def online_contrastive_loss(outputs, labels, loss_scale=None, num_items_in_batch
...
+    B       = logits.size(0)
+
+    # ----- teacher top-k data ------------------------------------------------
+    idx_list, lp_list = [], []
+    for p in logprobs_path:               # each file: (k, L_i)
+        idxs, lps = torch.load(p)
+        idx_list.append(idxs.T)           # (L_i, k)
+        lp_list.append(lps.T)
+    token_indices  = _pad_stack(idx_list,  pad_value=0)
+    token_logprobs = _pad_stack(lp_list, pad_value=-1e9)
```
Comment (young-agi) [inline swift/plugin/loss.py:417]: Make sure that this is in float32 type. Use `token_logprobs.float()`.

File: swift/plugin/loss.py (hunk 7 above)
Comment (Bezdarnost) [reply swift/plugin/loss.py:419]: it gave an error without int64, as I remember, torch asks for labels in int64

File: swift/plugin/loss.py (hunk 6 above)
Comment (Bezdarnost) [reply swift/plugin/loss.py]: I use in future teacher_logp, so I saved it
